mgr: add config to load modules in main interpreter instead of subinterpreter

athanatos opened the pull request:
<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

athanatos committed:
mgr: serialize python objects sent between subinterpreters via remote

Signed-off-by: Samuel Just <sjust@redhat.com>

athanatos committed:
python-common/.../service_spec: implement ServiceSpec.__getnewargs__ to allow unpickle to work correctly

Signed-off-by: Samuel Just <sjust@redhat.com>

athanatos committed:
mgr: add mgr_subinterpreter_modules config

This commit adds a mgr_subinterpreter_modules config to cause specified
modules (or all if * is specified) to be loaded in individual
subinterpreters.

This changes the default behavior of ceph-mgr from running each module
in a distinct subinterpreter to running them all in the same main
interpreter.  We can reintroduce subinterpreter support over time by
adding modules to the list as we test them.

Fixes: https://tracker.ceph.com/issues/73857
Fixes: https://tracker.ceph.com/issues/73859
Signed-off-by: Samuel Just <sjust@redhat.com>

athanatos commented:
Replaces https://github.com/ceph/ceph/pull/66176

idryomov commented:
I have just asked this in https://tracker.ceph.com/issues/73857, but may get more eyes here: what was the actual problem that sub-interpreters were supposed to solve?

athanatos commented:
Marking back to draft for now.  I see a potential problem with how MgrModule works -- MgrModule.COMMANDS seems like it'll end up with the command set for all modules in the same interpreter.  I need to look into this more on Monday, might just be misunderstanding how it works.

tchaikov commented:
@athanatos hi Sam, could you help take a look at  #66446 . now that python 3.13 (shipped by debian stable and other popular distros) provides dedicated APIs to manage subintepreters, probably we can also use them to address part of this problem?

athanatos committed:
mgr/Gil.cc: do not use PyGILState_Check()

See comment for explanation.

Fixes: https://tracker.ceph.com/issues/74220
Signed-off-by: Samuel Just <sjust@redhat.com>

athanatos committed:
mgr/Gil.cc: simplify Gil(), ~Gil()

Instead of restoring the passed ts and then swapping to
a fresh one, restore the fresh one in the first place.

Signed-off-by: Samuel Just <sjust@redhat.com>

NitzanMordhai reviewed (approved):
LGTM! will run teuthology test when we have a chance to get results from it

idryomov commented on src/common/options/mgr.yaml.in:
```diff
@@ -145,6 +145,19 @@ options:
+- name: mgr_subinterpreter_modules
+  type: str
+  level: advanced
+  desc: List of manager modules to load in independent submodules
```
```suggestion
  desc: List of manager modules to load in independent subinterpreters
```

batrick commented:
This PR is under test in [https://tracker.ceph.com/issues/74811](https://tracker.ceph.com/issues/74811).

tchaikov commented:
i think this changset also fixes https://tracker.ceph.com/issues/73930. 
the substantive change is commit 239b0dc8a9c4. but the referenced tracker tickets were

- https://tracker.ceph.com/issues/73857
- https://tracker.ceph.com/issues/73859
- https://tracker.ceph.com/issues/74220

i am updating the tracker ticket accordingly, to make sure its status is updated.
